### PR TITLE
Enable platform specific step of checking system logs

### PR DESCRIPTION
**Pull request by lizzha**
1. For each case, the current test controller collects the dmesg before and after the case run, and compare the results. This PR changes the step to be platform specific, so that it can be overwritten by each platform.
2. enable FailForLogCheck TestParameter, if one case set FailForLogCheck to True, then it will fail if log check fails
3. Test pass on Azure

### CODE REVIEW

File: Libraries/TestLogs.psm1 [hunk 1]
```diff
@@ -152,168 +152,72 @@ function Collect-CustomLogFile {
 	<#
 	.Synopsis
 		Checks if log file is present in VM and downloads it if so.
-    #>
-    param (
-        [string]$LogsDestination,
+	#>
```
Comment (juhlee-microsoft) [inline Libraries/TestLogs.psm1:146]: I suspect this indentation. Please check again.

File: Libraries/TestReport.psm1 [hunk 2]
```diff
@@ -252,6 +252,19 @@ Class JUnitReportGenerator
 	}
 }
 
+Class TestResult
```
Comment (juhlee-microsoft) [inline Libraries/TestReport.psm1]: Can you make the class name TestResults? Class and method are having the same name.

File: TestControllers/TestController.psm1 [hunk 3]
```diff
@@ -491,7 +502,12 @@ Class TestController
 			Write-LogInfo "==> Check if the test target machines are still running."
 			$isVmAlive = Is-VmAlive -AllVMDataObject $VMData -MaxRetryCount 10
 			if (!$global:IsWindowsImage -and $testParameters["SkipVerifyKernelLogs"] -ne "True" -and $isVmAlive -eq "True" ) {
-				GetAndCheck-KernelLogs -allDeployedVMs $VmData -status "Final" -EnableCodeCoverage $this.EnableCodeCoverage | Out-Null
+				$ret = $this.GetAndCompareOsLogs($VmData, "Final")
+				if ($testParameters["FailForLogCheck"] -eq "True" -and $ret -eq $false -and $currentTestResult.TestResult -eq $global:ResultPass) {
```
Comment (juhlee-microsoft) [inline TestControllers/TestController.psm1]: Please make it clear with extra parenthesis.

File: TestControllers/TestController.psm1 [hunk 4]
```diff
@@ -748,4 +764,95 @@ Class TestController
 			Write-LogErr "Source: Line $line in script $script_name."
 		}
 	}
+
+	[bool] GetAndCompareOsLogs($AllVMData, $Status) {
+		$retValue = $true
+		try	{
+			if (!($status -imatch "Initial" -or $status -imatch "Final")) {
+				Write-LogInfo "Status value should be either final or initial"
```
Comment (juhlee-microsoft) [inline TestControllers/TestController.psm1]: Write-LogErr ?

File: TestControllers/TestController.psm1 [hunk 5]
```diff
@@ -748,4 +764,95 @@ Class TestController
...
+						Write-LogErr "Failed to collect GCOV data from VM: $($VM.RoleName)"
+					}
+				}
+
+				Run-LinuxCmd -ip $VM.PublicIP -port $VM.SSHPort -runAsSudo `
+					-username $global:user -password $global:password `
+					-command "dmesg > ./${currentBootLogFile}" | Out-Null
+				Copy-RemoteFiles -download -downloadFrom $VM.PublicIP -port $VM.SSHPort -files "./${currentBootLogFile}" `
+					-downloadTo $BootLogDir -username $global:user -password $global:password | Out-Null
+				Write-LogInfo "$($VM.RoleName): $status Kernel logs collected successfully to ${currentBootLogFile} file."
```
Comment (juhlee-microsoft) [inline TestControllers/TestController.psm1]: Wording: $status kernel log, ${currentBootLogFile}, collected successfully.

File: TestControllers/TestController.psm1 [hunk 6]
```diff
@@ -748,4 +764,95 @@ Class TestController
...
+
+				Write-LogInfo "Checking for call traces in kernel logs.."
+				$KernelLogs = Get-Content $currentBootLog
+				$callTraceFound  = $false
+				foreach ($line in $KernelLogs) {
```
Comment (juhlee-microsoft) [inline TestControllers/TestController.psm1:812]: Suggestion: This foreach blocks are everywhere in the source code. Can you make it function and call? Code duplication.

**Comment by LiliDeng** [comment]
@lizzha Please run a test which has its own PS script, expected results pass, but it will return fail.
Before
```
Test Result: @{TestResult=PASS; TestSummary=; TestResultData=System.Object[]}
```
After
```
12/23/2019 09:06:46 : [INFO ] Test Result: TestResult
```

**Comment by lizzha** [comment]
> @lizzha Please run a test which has its own PS script, expected results pass, but it will return fail.
> Before
> Test Result: @{TestResult=PASS; TestSummary=; TestResultData=System.Object[]}
> 
> After
> 12/23/2019 09:06:46 : [INFO ] Test Result: TestResult

which test case is it?

File: Libraries/TestLogs.psm1 (hunk 1 above)
Comment (lizzha) [reply Libraries/TestLogs.psm1:146]: it was changed to tab, which is our convention

File: Libraries/TestReport.psm1 (hunk 2 above)
Comment (lizzha) [reply Libraries/TestReport.psm1]: Are you referring to the method in this class? If yes, it's the constructor and it must be the same name as the class

File: TestControllers/TestController.psm1 (hunk 3 above)
Comment (lizzha) [reply TestControllers/TestController.psm1]: will update

File: TestControllers/TestController.psm1 (hunk 4 above)
Comment (lizzha) [reply TestControllers/TestController.psm1]: Will update

File: TestControllers/TestController.psm1 (hunk 6 above)
Comment (lizzha) [reply TestControllers/TestController.psm1:812]: I suggest we file an issue for this, and file a separate PR

**Comment by LiliDeng** [comment]
> @lizzha Please run a test which has its own PS script, expected results pass, but it will return fail.
> Before
> Test Result: @{TestResult=PASS; TestSummary=; TestResultData=System.Object[]}
> After
> 12/23/2019 09:06:46 : [INFO ] Test Result: TestResult
> 
> which test case is it?

https://github.com/LIS/LISAv2/blob/master/Testscripts/Windows/SRIOV-RUN-COMMON-TEST.ps1#L161

**Comment by lizzha** [comment]
> @lizzha Please run a test which has its own PS script, expected results pass, but it will return fail.
> Before
> Test Result: @{TestResult=PASS; TestSummary=; TestResultData=System.Object[]}
> After
> 12/23/2019 09:06:46 : [INFO ] Test Result: TestResult
> which test case is it?
> 
> https://github.com/LIS/LISAv2/blob/master/Testscripts/Windows/SRIOV-RUN-COMMON-TEST.ps1#L161

I see. Some cases assume the return value of Collect-TestLogs is a string, but it's actually an object. When changing the testresult object to Class definition, it's more strict and these test scripts have an issue.

I've reverted the change for test result object for now, otherwise we need a full test pass to identify the test scripts that have the problem, which is low priority for now.

**Review by LiliDeng — APPROVED** [review]

**Pull request merged**